src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs

```diff
@@ -393,6 +397,7 @@ public override bool TryWrite(T item)
                     {
                         // The channel is full.  Just ignore the item being added
                         // but say we added it.
+                        _parent._itemDropped?.Invoke(item);
```

Comment:
We shouldn't invoke the delegate while the lock is held.  Same for all the case below.

Reply:
This is kinda tricky because I have to:
- store item that should be dropped outside of lock statement
- item is type of T which can be either struct/class, so I can not use default keyword here to signal whether some item should be dropped. I created helper struct that wraps item and has bool value signaling if struct has some item to be dropped.
- there are lot of return statements inside that lock block, which forced me to wrap it with try-finally and do invocation of delegate there

I am opened to suggestion, but currently this is best solution that came up to me.